File: app/src/main/java/vng/zalo/tdtai/zalo/zalo/viewmodels/ChatFragmentViewModel.java
package vng.zalo.tdtai.zalo.zalo.viewmodels;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.EventListener;
import com.google.firebase.firestore.FirebaseFirestoreException;
import com.google.firebase.firestore.QuerySnapshot;
import java.util.ArrayList;
import java.util.List;
import javax.annotation.Nullable;
import vng.zalo.tdtai.zalo.zalo.ZaloApplication;
import vng.zalo.tdtai.zalo.zalo.models.RoomItem;
import vng.zalo.tdtai.zalo.zalo.utils.Constants;

public class ChatFragmentViewModel extends ViewModel {
    private static final String TAG = ChatFragmentViewModel.class.getSimpleName();

    public MutableLiveData<List<RoomItem>> liveRoomItems;

    public ChatFragmentViewModel() {
        liveRoomItems = new MutableLiveData<>((List<RoomItem>) new ArrayList<RoomItem>());

        ZaloApplication.getFirebaseInstance()
                .collection(Constants.COLLECTION_USERS)
                .document(ZaloApplication.currentUserPhone)
                .collection(Constants.COLLECTION_ROOMS)
                .orderBy("lastMsgTime")
                .addSnapshotListener(new RoomEventListener());
    }

    class RoomEventListener implements EventListener<QuerySnapshot> {
        @Override
        public void onEvent(@Nullable QuerySnapshot queryDocumentSnapshots, @Nullable FirebaseFirestoreException e) {
            List<RoomItem> roomItems = new ArrayList<>();
            for(DocumentSnapshot doc: queryDocumentSnapshots){
                RoomItem roomItem = doc.toObject(RoomItem.class);
                roomItem.roomId = doc.getId();
                roomItems.add(roomItem);
            }
            liveRoomItems.setValue(roomItems);
        }
    }
}

File: app/src/main/java/vng/zalo/tdtai/zalo/zalo/viewmodels/ChatFragmentViewModel.kt
package vng.zalo.tdtai.zalo.zalo.viewmodels

import android.util.Log
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.google.firebase.firestore.ListenerRegistration
import com.google.firebase.firestore.Query
import kotlinx.coroutines.launch
import vng.zalo.tdtai.zalo.zalo.ZaloApplication
import vng.zalo.tdtai.zalo.zalo.models.RoomItem
import vng.zalo.tdtai.zalo.zalo.networks.Database

class ChatFragmentViewModel : ViewModel() {

    val liveRoomItems: MutableLiveData<List<RoomItem>> = MutableLiveData(ArrayList())
    var listenerRegistrations = ArrayList<ListenerRegistration>()

    init {
        viewModelScope.launch {

        }
        val userRoomListener = Database.addUserRoomsListener(
                userPhone = ZaloApplication.currentUser!!.phone!!,
                fieldToOrder = "lastMsgTime",
                orderDirection = Query.Direction.DESCENDING
        ) { liveRoomItems.value = it }
        listenerRegistrations.add(userRoomListener)
    }

    fun removeListeners() {
        listenerRegistrations.forEach { it.remove() }
        Log.d("removeListeners", "finalize")
    }
}